metis-media-service/src/main/java/eu/europeana/metis/mediaprocessing/http/AbstractHttpClient.java

```diff
@@ -106,71 +79,139 @@ private void cleanConnections(PoolingHttpClientConnectionManager connectionManag
   public R download(I resourceEntry) throws IOException {
 
     // Set up the connection.
-    final String resourceUlr = getResourceUrl(resourceEntry);
-    final HttpGet httpGet = new HttpGet(resourceUlr);
-    final HttpClientContext context = HttpClientContext.create();
+    final URI resourceUlr = URI.create(getResourceUrl(resourceEntry));
+
+    BodyHandler<InputStream> handler = BodyHandlers.ofInputStream();
+    CancelableBodyWrapper<InputStream> bodyWrapper = new CancelableBodyWrapper<>(handler);
+
+     httpResponse = makeHttpRequest(resourceUlr, bodyWrapper);
 
     // Set up the abort trigger
     final TimerTask abortTask = new TimerTask() {
       @Override
       public void run() {
-        LOGGER.info("Aborting request due to time limit: {}.", resourceUlr);
-        httpGet.abort();
+        LOGGER.info("Aborting request due to time limit: {}.", resourceUlr.getPath());
+        bodyWrapper.cancel();
+        if (httpResponse.body() != null) {
+          try {
+            httpResponse.body().close();
+          } catch (IOException e) {
+            LOGGER.warn(
+                    "Something went wrong while trying to close the input stream after cancelling the http request.",
+                    e);
+          }
+        }
       }
     };
     final Timer timer = new Timer(true);
     timer.schedule(abortTask, requestTimeout);
 
-    // Execute the request.
-    try (final CloseableHttpResponse responseObject = client.execute(httpGet, context)) {
-
-      // Do first analysis
-      final HttpEntity responseEntity = performThrowingFunction(responseObject, response -> {
-        final int status = response.getCode();
-        if (!httpCallIsSuccessful(status)) {
-          throw new IOException("Download failed of resource " + resourceUlr + ". Status code " +
-                  status + " (message: " + response.getReasonPhrase() + ").");
-        }
-        return response.getEntity();
-      });
+    final int statusCode;
+
+    if (httpResponse != null) {
+      statusCode = httpResponse.statusCode();
+    } else {
+      throw new IOException("A problem occurred when sending the request");
+    }
+
+    final Optional<String> redirectUris = httpResponse.headers().firstValue("Location");
+    final URI actualUri;
+
+    // Do first check redirection and analysis
+    if(Family.familyOf(statusCode) == Family.REDIRECTION){
+      httpResponse = performRedirect(statusCode, resourceUlr.resolve(redirectUris.get()),
```

Comment:
httpResponse according to performRedirect can return null, although it's not possible in this case.
Nonetheless nullity should be also checked here, or performRedirect shouldn't have the possibility to return null.

Reply:
Hmm ok, I'll make the method `performRedirct` throw an exception in case the response is still `null`